client/app/components/dashboards/add-widget-dialog.html

```diff
@@ -4,46 +4,75 @@ <h4 class="modal-title">Add Widget</h4>
 </div>
 <div class="modal-body">
   <p class="btn-group">
-      <button type="button" class="btn btn-default" ng-class="{active: $ctrl.isVisualization()}" ng-click="$ctrl.setType('visualization')">Visualization</button>
-      <button type="button" class="btn btn-default" ng-class="{active: $ctrl.isTextBox()}" ng-click="$ctrl.setType('textbox')">Text Box</button>
+    <button type="button" class="btn btn-default" ng-class="{active: $ctrl.isVisualization}" ng-click="$ctrl.setType('visualization')">Visualization</button>
+    <button type="button" class="btn btn-default" ng-class="{active: $ctrl.isTextBox}" ng-click="$ctrl.setType('textbox')">Text Box</button>
   </p>
 
-  <div ng-show="$ctrl.isTextBox()">
-      <div class="form-group">
-          <textarea class="form-control" ng-model="$ctrl.text" rows="3"></textarea>
+  <div ng-show="$ctrl.isTextBox">
+    <div class="form-group">
+      <textarea class="form-control" ng-model="$ctrl.text" ng-model-options="{ debounce: 200 }" rows="3"></textarea>
+    </div>
+    <div ng-show="$ctrl.text">
+      <strong>Preview:</strong>
+      <p ng-bind-html="$ctrl.text | markdown"></p>
+    </div>
+  </div>
+
+  <div ng-show="$ctrl.isVisualization">
+    <div class="form-group">
+      <input type="text" placeholder="Search a query by name" class="form-control" autofocus="autofocus"
+        ng-if="!$ctrl.selectedQuery" ng-change="$ctrl.searchQueries($ctrl.searchTerm)"
+        ng-model="$ctrl.searchTerm" ng-model-options="{ debounce: 200 }">
+      <div ng-if="$ctrl.selectedQuery" class="p-relative">
+        <input type="text" class="form-control bg-white"
+          ng-value="$ctrl.selectedQuery.name" readonly="readonly">
+        <a href="javascript:void(0)" ng-click="$ctrl.selectQuery(null)"
+          class="d-flex align-items-center justify-content-center"
+          style="position: absolute; right: 1px; top: 1px; bottom: 1px; width: 30px; background: #fff; border-radius: 3px;"
+        ><i class="text-muted fa fa-times"></i></a>
       </div>
-      <div ng-show="$ctrl.text">
-          <strong>Preview:</strong>
-          <p ng-bind-html="$ctrl.text | markdown"></p>
+    </div>
+
+    <div ng-if="!$ctrl.selectedQuery && ($ctrl.searchTerm == '')" class="scrollbox" style="max-height: 50vh">
+      <div class="list-group" ng-if="$ctrl.recentQueries.length > 0">
+        <a href="javascript:void(0)" class="list-group-item"
```

Comment:
Are you sure the `javascript:void(0);" part needed?

Reply:
yes, otherwise links will not have their style (color, underline, etc.)

Reply:
Here we have similar markup and I think it's ok, no?

https://github.com/getredash/redash/blob/17f3b1e92df35613677584958cc37e7aee783990/client/app/pages/dashboards/dashboard-list.html#L9-L12

Reply:
👍

Reply:
ah, there is s small difference between links with and without `href`: link without `href` will have cursor like for regular text, but link with `href` will have `cursor: pointer` - it's defaults for most browsers. it will work also for `list-group`s (so I'll remove `href`s here) because bootstrap implicitly sets cursor for list items.

there also was a bug in old IE (up to IE8, I think) where links without `href` didn't have usual hover styling, but it's irrelevant for us.